man/theme.conf.5
.TH WMI 5 User Manuals
.SH NAME
WMI \- Window Manager Improved, a X11 window manager for hackers
.SH DESCRIPTION
The file \fItheme.conf\f1 is used to define the look and feel of WMI.
.SH FORMAT
The basic format of all \fI.conf\f1 and \fI.session\f1 files is as follows:

key=value

# comment

key=\\#value # '\\' escapes special characters

# special characters are: ", =, \\ and #

If you've whitespaces within values, you've to quote them with '"' characters. You're free to let empty lines or whitespaces before and after keys, the '=' operator and the value. If you need a special character, you've to escape it using the '\\' character, e.g. \\=.
.SH COLOR FORMAT
Note that you can vary the syntax for color triplet definitions as follows:

key=\\#nnnnnn

or

key="#nnnnnn"

or

key=blue

This shows different escaping possibilities, the last case is useful for predefined X11 colors.

The syntax: key=rgb:/nn/nn/nn
is NOT supported.
.SH TERMINOLOGY
All color definitions begin with prefix 'color.', afterwards the object is appended to this prefix, e.g. 'label', 'button'.

There are four types of colors per object (note that some have only three colors): 'background', 'text', 'shine' and 'shadow'. Background means the background filling color of the object, text means the color of text, if any, shine and shadow means the color of the objects border, where top-left border is the shine border and bottom-right is the shadow border.

If an object is distinguished into active and inactive it may have the focus, like frames or tabs.
.SH FORMAT
Following look and feel option keys are defined:
.SH OPTIONS
.TP
\fBcolor.bar.background=\f1
Background color of float and status bar and input mode. 
.TP
\fBcolor.bar.shadow=\f1
Shadow border of the float and status bar. 
.TP
\fBcolor.bar.shine=\f1
Shine border of the float and status bar. 
.TP
\fBcolor.bar.text=\f1
Text color of the status text within status bar and the input mode prompt. 
.TP
\fBcolor.button.background.normal=\f1
Background color of not pressed buttons. 
.TP
\fBcolor.button.background.pressed=\f1
Background color of pressed buttons. 
.TP
\fBcolor.button.border.shadow.normal=\f1
Border shadow color of not pressed buttons. 
.TP
\fBcolor.button.border.shadow.pressed=\f1
Border shadow color of pressed buttons. 
.TP
\fBcolor.button.border.shine.normal=\f1
Border shine color of not pressed buttons. 
.TP
\fBcolor.button.border.shine.pressed=\f1
Border shine color of pressed buttons. 
.TP
\fBcolor.button.figure.shadow.normal=\f1
Figure shadow color of not pressed buttons. 
.TP
\fBcolor.button.figure.shadow.pressed=\f1
Figure shadow color of pressed buttons. 
.TP
\fBcolor.button.figure.shine.normal=\f1
Figure shine color of not pressed buttons. 
.TP
\fBcolor.focusreq.background=\f1
Background color of an item which requests for focus. 
.TP
\fBcolor.focusreq.text=\f1
Text color of an item which requests for focus. 
.TP
\fBcolor.focusreq.shine=\f1
Shine border color of an item which requests for focus. 
.TP
\fBcolor.focusreq.shadow=\f1
Shadow border color of an item which requests for focus. 
.TP
\fBcolor.button.figure.shine.pressed=\f1
Figure shine color of pressed buttons. 
.TP
\fBcolor.frame.background.focussed=\f1
Background color of the focussed frame. 
.TP
\fBcolor.frame.background.normal=\f1
Background color of not focussed frames. 
.TP
\fBcolor.frame.pseudo=\f1
XORed color of frame/client movements. 
.TP
\fBcolor.frame.shadow.focussed=\f1
Shadow color of the focussed frame. 
.TP
\fBcolor.frame.shadow.normal=\f1
Shadow color of not focussed frames. 
.TP
\fBcolor.frame.shine.focussed=\f1
Shine color of the focussed frame. 
.TP
\fBcolor.frame.shine.normal=\f1
Shine color of not focussed frames. 
.TP
\fBcolor.label.background.focussed=\f1
Background color of focussed labels (workspace and group pager labels or focussed entry within input mode). 
.TP
\fBcolor.label.background.normal=\f1
Background color of not focussed labels (workspace and group pager labels or focussed entry within input mode). 
.TP
\fBcolor.label.shadow.focussed=\f1
Shadow border color of focussed labels (workspace and group pager labels or focussed entry within input mode). 
.TP
\fBcolor.label.shadow.normal=\f1
Shadow border color of not focussed labels (workspace and group pager labels or focussed entry within input mode). 
.TP
\fBcolor.label.shine.focussed=\f1
Shine border color of focussed labels (workspace and group pager labels or focussed entry within input mode). 
.TP
\fBcolor.label.shine.normal=\f1
Shine border color of not focussed labels (workspace and group pager labels or focussed entry within input mode). 
.TP
\fBcolor.label.text.focussed=\f1
Text color of focussed labels (workspace and group pager labels or focussed entry within input mode). 
.TP
\fBcolor.label.text.normal=\f1
Text color of not focussed labels (workspace and group pager labels or focussed entry within input mode). 
.TP
\fBcolor.meter.background=\f1
Background color of a meter. 
.TP
\fBcolor.meter.figure.high=\f1
High color of a meter. 
.TP
\fBcolor.meter.figure.normal=\f1
Normal color of a meter. 
.TP
\fBcolor.meter.figure.low=\f1
Low color of a meter. 
.TP
\fBcolor.meter.border.shine=\f1
Shine border color of a meter. 
.TP
\fBcolor.meter.border.shadow=\f1
Shadow border color of a meter. 
.TP
\fBcolor.meter.text=\f1
Text color of a meter. 
.TP
\fBcolor.tab.background.active.focussed=\f1
Background color of active tab within focussed frame. 
.TP
\fBcolor.tab.background.active.normal=\f1
Background color of active tab within not focussed frame. 
.TP
\fBcolor.tab.background.inactive.focussed=\f1
Background color of inactive tab within focussed frame. 
.TP
\fBcolor.tab.background.inactive.normal=\f1
Background color of inactive tab within not focussed frame. 
.TP
\fBcolor.tab.shadow.active.focussed=\f1
Shadow color of active tab within focussed frame. 
.TP
\fBcolor.tab.shadow.active.normal=\f1
Shadow color of active tab within not focussed frame. 
.TP
\fBcolor.tab.shadow.inactive.focussed=\f1
Shadow color of inactive tab within focussed frame. 
.TP
\fBcolor.tab.shadow.inactive.normal=\f1
Shadow color of inactive tab within not focussed frame. 
.TP
\fBcolor.tab.shine.active.focussed=\f1
Shine color of active tab within focussed frame. 
.TP
\fBcolor.tab.shine.active.normal=\f1
Shine color of active tab within not focussed frame. 
.TP
\fBcolor.tab.shine.inactive.focussed=\f1
Shine color of inactive tab within focussed frame. 
.TP
\fBcolor.tab.shine.inactive.normal=\f1
Shine color of inactive tab within not focussed frame. 
.TP
\fBcolor.tab.text.active.focussed=\f1
Text color of active tab within focussed frame. 
.TP
\fBcolor.tab.text.active.normal=\f1
Text color of active tab within not focussed frame. 
.TP
\fBcolor.tab.text.inactive.focussed=\f1
Text color of inactive tab within focussed frame. 
.TP
\fBcolor.tab.text.inactive.normal=\f1
Text color of inactive tab within not focussed frame. 
.TP
\fBcolor.tiled.shine.focussed=\f1
Shine border color of focussed tiled clients. 
.TP
\fBcolor.tiled.shadow.focussed=\f1
Shadow border color of focussed tiled clients. 
.TP
\fBcolor.tiled.shine.normal=\f1
Shine border color of unfocussed tiled clients. 
.TP
\fBcolor.tiled.shadow.normal=\f1
Shine border color of unfocussed tiled clients. 
.TP
\fBexec=\f1
Execution of a command, e.g. setting the background wallpaper which is invoked while loading the theme (while startup of WMI). 
.TP
\fBfont=\f1
The font of labels, tabs and bars. Note that the height of buttons and bars is scaled to match with the font size of the defined font. If you want to have bigger or smaller bars and buttons simply choose a bigger or smaller font.
.SH AUTHOR
WMI was written by Anselm R. Garbe < anselm@garbe.us > and others (see AUTHORS or the WMI site for further information)

You will find the newest version of WMI at \fBhttp://wmi.suckless.org/\f1. There is also a mailing list.
.SH BUGS
You should report them to the mailing list.
.SH SEE ALSO
\fBwmi(1)\f1, \fBwmiremote(1)\f1, \fBactions.conf(5)\f1, \fBcommon.conf(5)\f1 \fBsession.conf(5)\f1, 


NEWS
$ Id: $
Visit http://wmi.berlios.de for current news about the WMI.


theme.cpp
// Copyright (c) 2003 - 2009 Anselm R Garbe <anselm@garbe.us>
// See LICENSE for license details.

#include "theme.h"

#include "logger.h"
#include "monitor.h"
#include "xcore.h"

Theme::Theme() {
}

Theme::~Theme() {
}

unsigned long Theme::allocateColor(string name, Colormap cmap) {

    XColor color;

    if (!XCORE->allocNamedColor(cmap, name, &color))
    {
        Logger::instance()->error("cannot allocate color '"
                                   + name + "'", true);
        // exits WMI
    }

    return color.pixel;
}

void Theme::initTheme(MSettings *ts, Colormap cmap) {

    BAR_BACKGROUND = allocateColor(Util::get(ts, "color.bar.background"), cmap);
    BAR_TEXT = allocateColor(Util::get(ts, "color.bar.text"), cmap);
    BAR_SHINE = allocateColor(Util::get(ts, "color.bar.shine"), cmap);
    BAR_SHADOW = allocateColor(Util::get(ts, "color.bar.shadow"), cmap);

    METER_BACKGROUND = allocateColor(Util::get(ts, "color.meter.background"), cmap);
    METER_FIGURE_HIGH = allocateColor(Util::get(ts, "color.meter.figure.high"), cmap);
    METER_FIGURE_NORMAL = allocateColor(Util::get(ts, "color.meter.figure.normal"), cmap);
    METER_FIGURE_LOW = allocateColor(Util::get(ts, "color.meter.figure.low"), cmap);
    METER_BORDER_SHINE= allocateColor(Util::get(ts, "color.meter.border.shine"), cmap);
    METER_BORDER_SHADOW= allocateColor(Util::get(ts, "color.meter.border.shadow"), cmap);

    FOCUSREQ_BACKGROUND = allocateColor(Util::get(ts, "color.focusreq.background"), cmap);
    FOCUSREQ_TEXT = allocateColor(Util::get(ts, "color.focusreq.text"), cmap);
    FOCUSREQ_SHINE = allocateColor(Util::get(ts, "color.focusreq.shine"), cmap);
    FOCUSREQ_SHADOW = allocateColor(Util::get(ts, "color.focusreq.shadow"), cmap);

    TILED_SHINE_FOCUSSED = allocateColor(Util::get(ts, "color.tiled.shine.focussed"), cmap);
    TILED_SHADOW_FOCUSSED = allocateColor(Util::get(ts, "color.tiled.shadow.focussed"), cmap);
    TILED_SHINE_NORMAL = allocateColor(Util::get(ts, "color.tiled.shine.normal"), cmap);
    TILED_SHADOW_NORMAL = allocateColor(Util::get(ts, "color.tiled.shadow.normal"), cmap);

    BUTTON_BACKGROUND_NORMAL =
        allocateColor(Util::get(ts, "color.button.background.normal"), cmap);
    BUTTON_BACKGROUND_PRESSED =
        allocateColor(Util::get(ts, "color.button.background.pressed"), cmap);
    BUTTON_SHINE_BORDER_NORMAL =
        allocateColor(Util::get(ts, "color.button.border.shine.normal"), cmap);
    BUTTON_SHINE_BORDER_PRESSED =
        allocateColor(Util::get(ts, "color.button.border.shine.pressed"), cmap);
    BUTTON_SHADOW_BORDER_NORMAL =
        allocateColor(Util::get(ts, "color.button.border.shadow.normal"), cmap);
    BUTTON_SHADOW_BORDER_PRESSED =
        allocateColor(Util::get(ts, "color.button.border.shadow.pressed"), cmap);
    BUTTON_SHINE_FIGURE_NORMAL =
        allocateColor(Util::get(ts, "color.button.figure.shine.normal"), cmap);
    BUTTON_SHINE_FIGURE_PRESSED =
        allocateColor(Util::get(ts, "color.button.figure.shine.pressed"), cmap);
    BUTTON_SHADOW_FIGURE_NORMAL =
        allocateColor(Util::get(ts, "color.button.figure.shadow.normal"), cmap);
    BUTTON_SHADOW_FIGURE_PRESSED =
        allocateColor(Util::get(ts, "color.button.figure.shadow.pressed"), cmap);

    FRAME_BACKGROUND_NORMAL =
        allocateColor(Util::get(ts, "color.frame.background.normal"), cmap);
    FRAME_BACKGROUND_FOCUSSED =
        allocateColor(Util::get(ts, "color.frame.background.focussed"), cmap);
    FRAME_SHINE_NORMAL =
        allocateColor(Util::get(ts, "color.frame.shine.normal"), cmap);
    FRAME_SHINE_FOCUSSED =
        allocateColor(Util::get(ts, "color.frame.shine.focussed"), cmap);
    FRAME_SHADOW_NORMAL =
        allocateColor(Util::get(ts, "color.frame.shadow.normal"), cmap);
    FRAME_SHADOW_FOCUSSED =
        allocateColor(Util::get(ts, "color.frame.shadow.focussed"), cmap);
    FRAME_PSEUDO =
        allocateColor(Util::get(ts, "color.frame.pseudo"), cmap);

    LABEL_BACKGROUND_NORMAL =
        allocateColor(Util::get(ts, "color.label.background.normal"), cmap);
    LABEL_BACKGROUND_FOCUSSED =
        allocateColor(Util::get(ts, "color.label.background.focussed"), cmap);
    LABEL_TEXT_NORMAL =
        allocateColor(Util::get(ts, "color.label.text.normal"), cmap);
    LABEL_TEXT_FOCUSSED =
        allocateColor(Util::get(ts, "color.label.text.focussed"), cmap);
    LABEL_SHINE_NORMAL =
        allocateColor(Util::get(ts, "color.label.shine.normal"), cmap);
    LABEL_SHINE_FOCUSSED =
        allocateColor(Util::get(ts, "color.label.shine.focussed"), cmap);
    LABEL_SHADOW_NORMAL =
        allocateColor(Util::get(ts, "color.label.shadow.normal"), cmap);
    LABEL_SHADOW_FOCUSSED =
        allocateColor(Util::get(ts, "color.label.shadow.focussed"), cmap);

    TAB_BACKGROUND_ACTIVE_NORMAL =
        allocateColor(Util::get(ts, "color.tab.background.active.normal"), cmap);
    TAB_BACKGROUND_INACTIVE_NORMAL =
        allocateColor(Util::get(ts, "color.tab.background.inactive.normal"), cmap);
    TAB_BACKGROUND_ACTIVE_FOCUSSED =
        allocateColor(Util::get(ts, "color.tab.background.active.focussed"), cmap);
    TAB_BACKGROUND_INACTIVE_FOCUSSED =
        allocateColor(Util::get(ts, "color.tab.background.inactive.focussed"), cmap);
    TAB_TEXT_ACTIVE_NORMAL =
        allocateColor(Util::get(ts, "color.tab.text.active.normal"), cmap);
    TAB_TEXT_INACTIVE_NORMAL =
        allocateColor(Util::get(ts, "color.tab.text.inactive.normal"), cmap);
    TAB_TEXT_ACTIVE_FOCUSSED =
        allocateColor(Util::get(ts, "color.tab.text.active.focussed"), cmap);
    TAB_TEXT_INACTIVE_FOCUSSED =
        allocateColor(Util::get(ts, "color.tab.text.inactive.focussed"), cmap);
    TAB_SHINE_ACTIVE_NORMAL =
        allocateColor(Util::get(ts, "color.tab.shine.active.normal"), cmap);
    TAB_SHINE_INACTIVE_NORMAL =
        allocateColor(Util::get(ts, "color.tab.shine.inactive.normal"), cmap);
    TAB_SHINE_ACTIVE_FOCUSSED =
        allocateColor(Util::get(ts, "color.tab.shine.active.focussed"), cmap);
    TAB_SHINE_INACTIVE_FOCUSSED =
        allocateColor(Util::get(ts, "color.tab.shine.inactive.focussed"), cmap);
    TAB_SHADOW_ACTIVE_NORMAL =
        allocateColor(Util::get(ts, "color.tab.shadow.active.normal"), cmap);
    TAB_SHADOW_INACTIVE_NORMAL =
        allocateColor(Util::get(ts, "color.tab.shadow.inactive.normal"), cmap);
    TAB_SHADOW_ACTIVE_FOCUSSED =
        allocateColor(Util::get(ts, "color.tab.shadow.active.focussed"), cmap);
    TAB_SHADOW_INACTIVE_FOCUSSED =
        allocateColor(Util::get(ts, "color.tab.shadow.inactive.focussed"), cmap);
}
